streamer/test/unit/streamer/test_cache.py

```diff
@@ -0,0 +1,105 @@
+from unittest import TestCase
+
+from mock import Mock, patch
+
+from pulp.streamer.cache import Cache, Item, NotCached
+
+MODULE = 'pulp.streamer.cache'
+
+
+class TestCache(TestCase):
+
+    def test_basic_operation(self):
+        t1 = Mock()
+        t2 = Mock()
+        cache = Cache()
+        cache.add('t1', t1)
+        cache.add('t2', t2)
+        for n in range(10):
+            self.assertEqual(t1, cache.get('t1'))
+        for n in range(10):
+            self.assertEqual(t2, cache.get('t2'))
+
+    @patch(MODULE + '.Item.now')
+    def test_add(self, now):
+        t1 = Mock()
+        cache = Cache()
+        cache.add('t1', t1)
+        item = cache._inventory['t1']
+        self.assertEqual(item.object, t1)
+        self.assertEqual(item.ref_count, 1)
+        self.assertEqual(item.last_requested, now.return_value)
+
+    def test_purge(self):
+        t1 = Mock()
+        cache = Cache()
+        cache.add('t1', t1)
+        self.assertTrue('t1' in cache._inventory)
+        cache.purge('t1')
+        self.assertFalse('t1' in cache._inventory)
+
+    @patch(MODULE + '.Item.now')
+    def test_get(self, now):
+        t1 = Mock()
+        now.side_effect = [1, 2, 3]
+        key = 't1'
+        cache = Cache()
+        cache.add(key, t1)
+        gotten = cache.get(key)
+        self.assertEqual(cache._inventory[key].last_requested, 2)
+        self.assertEqual(gotten, t1)
+        self.assertRaises(NotCached, cache.get, 'xx')
+
+    @patch(MODULE + '.Item.now')
+    def test_evict(self, now):
+        now.side_effect = [1, 2, 3, 4]
+        key = 't1'
+        cache = Cache()
+        cache.add(key, Mock(key=key))
+        cache.eviction_threshold = 3
+        evicted = cache.evict()
+        self.assertTrue('t1' in cache)
+        self.assertEqual(evicted, [])
+        evicted = cache.evict()
+        self.assertTrue('t1' in cache)
+        self.assertEqual(evicted, [])
+        evicted = cache.evict()
+        self.assertFalse('t1' in cache)
+        self.assertEqual([obj.key for obj in evicted], [key])
+
+    @patch(MODULE + '.Item.now')
+    def test_evict_busy(self, now):
+        now.side_effect = [1, 2, 3, 4]
+        key = 't1'
+        t1 = Mock()  # hold ref to make it busy.
+        cache = Cache()
+        cache.add(key, t1)
+        cache.eviction_threshold = 0
+        cache.evict()
+        self.assertTrue('t1' in cache)
```

Comment:
If this is an intended use case, it's worth implementing ``__contains__``:

https://docs.python.org/2/reference/datamodel.html#object.__contains__

As-is, this line would cause an iteration through all of the keys in the cache until ``t1`` is found. Not a big deal with small collections, but can make a huge difference as size increases.